fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/MetaDataProtoEditor.java

```diff
@@ -0,0 +1,234 @@
+            throw new MetaDataException("Use addNestedRecordType for adding NESTED record types");
+        }
+        if (findMessageTypeByName(metaDataBuilder.getRecordsBuilder(), newRecordType.getName()) != null) {
+            throw new MetaDataException("Record type " + newRecordType.getName() + " already exists");
+        }
+        metaDataBuilder.getRecordsBuilder().addMessageType(newRecordType);
+        addFieldToUnion(metaDataBuilder.getRecordsBuilder(), newRecordType);
+    }
+
+    private static void addFieldToUnion(@Nonnull DescriptorProtos.FileDescriptorProto.Builder fileBuilder, @Nonnull DescriptorProtos.DescriptorProto newRecordType) {
+        DescriptorProtos.DescriptorProto.Builder unionBuilder = fetchUnionBuilder(fileBuilder);
+        if (unionBuilder.getOneofDeclCount() > 0) {
+            throw new MetaDataException("Adding record type to oneOf is not allowed");
+        }
+        DescriptorProtos.FieldDescriptorProto.Builder fieldBuilder = DescriptorProtos.FieldDescriptorProto.newBuilder()
+                .setLabel(DescriptorProtos.FieldDescriptorProto.Label.LABEL_OPTIONAL)
+                .setType(DescriptorProtos.FieldDescriptorProto.Type.TYPE_MESSAGE)
+                .setTypeName(newRecordType.getName())
+                .setName("_" + newRecordType.getName())
+                .setNumber(assignFieldNumber(unionBuilder));
+        unionBuilder.addField(fieldBuilder);
+    }
+
+    @Nonnull
+    private static DescriptorProtos.DescriptorProto.Builder fetchUnionBuilder(@Nonnull DescriptorProtos.FileDescriptorProto.Builder fileBuilder) {
+        for (DescriptorProtos.DescriptorProto.Builder messageTypeBuilder : fileBuilder.getMessageTypeBuilderList()) {
+            if (isUnion(messageTypeBuilder)) {
+                return messageTypeBuilder;
+            }
+        }
+        throw new MetaDataException("Union descriptor not found");
+    }
+
+    private static boolean isUnion(@Nonnull DescriptorProtos.DescriptorProto.Builder messageTypeBuilder) {
+        if (messageTypeBuilder.getName().equals(RecordMetaDataBuilder.DEFAULT_UNION_NAME)) {
+            return true;
+        }
+        RecordMetaDataOptionsProto.RecordTypeOptions recordTypeOptions = messageTypeBuilder.getOptions()
+                .getExtension(RecordMetaDataOptionsProto.record);
+        return recordTypeOptions != null &&
+               recordTypeOptions.hasUsage() &&
+               recordTypeOptions.getUsage() == RecordMetaDataOptionsProto.RecordTypeOptions.Usage.UNION;
+    }
+
+    private static int assignFieldNumber(@Nonnull DescriptorProtos.DescriptorProto.Builder messageType) {
+        if (messageType.getFieldCount() == 0) {
+            return 1;
+        }
+        return messageType.getFieldList().stream().mapToInt(DescriptorProtos.FieldDescriptorProto::getNumber).max().getAsInt() + 1;
+    }
+
+    /**
+     * Add a new {@code NESTED} record type to the meta-data, which can be used to define fields in other record types.
```

Comment:
```suggestion
     * Add a new {@code NESTED} record type to the meta-data. This can be used to define fields in other record types, but it does not add the new record type to the union.
```